L1Trigger/TrackerTFP/src/KalmanFilter.cc

```diff
@@ -0,0 +1,398 @@
+      const int channelTrack = offset + channel;
+      const StreamTrack& streamTracks = streamsTrack[channelTrack];
+      vector<TrackKFin*>& tracks = input_[channel];
+      tracks.reserve(streamTracks.size());
+      for (int frame = 0; frame < (int)streamTracks.size(); frame++) {
+        const FrameTrack& frameTrack = streamTracks[frame];
+        if (frameTrack.first.isNull())
+          continue;
+        const auto end = find_if(next(streamTracks.begin(), frame + 1), streamTracks.end(), valid);
+        const int size = distance(next(streamTracks.begin(), frame), end);
+        tracks.insert(tracks.end(), size - 1, nullptr);
+        deque<StubKFin*> stubs;
+        for (int layer = 0; layer < setup_->numLayers(); layer++) {
+          const int channelStub = channelTrack * setup_->numLayers() + layer;
+          const TTDTC::Stream& streamStubs = streamsStub[channelStub];
+          for (int i = frame; i < frame + size; i++) {
+            const TTDTC::Frame& frameStub = streamStubs[i];
+            if (frameStub.first.isNull())
+              break;
+            stubs_.emplace_back(frameStub, dataFormats_, layer);
+            stubs.push_back(&stubs_.back());
+          }
+        }
+        tracks_.emplace_back(frameTrack, dataFormats_, vector<StubKFin*>(stubs.begin(), stubs.end()));
+        /*cout << "KFin " << tracks_.back().inv2R() << " " << tracks_.back().phiT() << endl;
+        cout << "KFin " << -.5 * frameTrack.first->rInv() << endl;*/
+        tracks.push_back(&tracks_.back());
+      }
+    }
+  }
+
+  // fill output products
+  void KalmanFilter::produce(TTDTC::Streams& acceptedStubs, StreamsTrack& acceptedTracks, TTDTC::Streams& lostStubs, StreamsTrack& lostTracks) {
+    auto put = [this](const deque<State*>& states, TTDTC::Streams& streamsStubs, StreamsTrack& streamsTracks, int channel) {
+      const int streamId = region_ * dataFormats_->numChannel(Process::kf) + channel;
+      const int offset = streamId * setup_->numLayers();
+      StreamTrack& tracks = streamsTracks[streamId];
+      tracks.reserve(states.size());
+      for (int layer = 0; layer < setup_->numLayers(); layer++)
+        streamsStubs[offset + layer].reserve(states.size());
+      for (State* state : states) {
+        tracks.emplace_back(state->frame());
+        for (const StubKF& stub : state->stubs())
+          streamsStubs[offset + stub.layer()].emplace_back(stub.frame());
+        // adding a gap to all layer without a stub
+        for (int layer : state->hitPattern().ids(false))
+          streamsStubs[offset + layer].emplace_back(TTDTC::Frame());
+      }
+    };
+    for (int channel = 0; channel < dataFormats_->numChannel(Process::kf); channel++) {
+      deque<State*> stream;
+      deque<State*> lost;
+      // proto state creation
+      int trackId(0);
+      for (TrackKFin* track : input_[channel]) {
+        State* state = nullptr;
+        if (track) {
+          states_.emplace_back(dataFormats_, track, trackId++);
```

Comment:
Add comment "Store states, so remainder of KF algo can work with pointers to them (saves CPU)".

Reply:
done